Mathlib/Analysis/CStarAlgebra/ContinuousFunctionalCalculus/Range.lean

```diff
@@ -0,0 +1,296 @@
+  generated by the given element.
+* `Commute.cfc` and `Commute.cfcₙ`: an element commutes with `cfc f a` or `cfcₙ f a` if it
+  commutes with both `a` and `star a`. Specialized versions for `ℝ` and `ℝ≥0` or for
+  `IsSelfAdjoint a` which do not require the user to show the element commutes with `star a` are
+  provided for convenience.
+
+-/
+
+open Topology
+
+open scoped CStarAlgebra
+
+section CFCRangeCommute
+
+theorem range_cfc (R : Type*) {A : Type*} {p : A → Prop} [CommSemiring R] [StarRing R]
+    [MetricSpace R] [IsTopologicalSemiring R] [ContinuousStar R] [Ring A] [StarRing A]
+    [Algebra R A] [TopologicalSpace A] [StarModule R A] [ContinuousFunctionalCalculus R A p]
+    {a : A} (ha : p a) : Set.range (cfc (R := R) · a) = (cfcHom ha (R := R)).range := by
+  ext
+  constructor
+  all_goals rintro ⟨f, rfl⟩
+  · exact cfc_cases _ a f (zero_mem _) fun hf ha ↦ ⟨_, rfl⟩
+  · exact ⟨Subtype.val.extend f 0, cfcHom_eq_cfc_extend _ ha _ |>.symm⟩
+
+section RCLike
+
+variable (𝕜 : Type*) {A : Type*} {p : A → Prop} [RCLike 𝕜] [Ring A] [StarRing A] [Algebra 𝕜 A]
+variable [TopologicalSpace A] [StarModule 𝕜 A] [ContinuousFunctionalCalculus 𝕜 A p]
+variable [IsTopologicalRing A] [ContinuousStar A]
+
+open StarAlgebra
+
+open scoped ContinuousFunctionalCalculus in
+theorem range_cfcHom {a : A} (ha : p a) :
+    (cfcHom ha (R := 𝕜)).range = elemental 𝕜 a := by
+  rw [StarAlgHom.range_eq_map_top, ← ContinuousMap.elemental_id_eq_top, StarAlgebra.elemental,
+    ← StarSubalgebra.topologicalClosure_map _ _ (cfcHom_isClosedEmbedding ha (R := 𝕜)).isClosedMap
+      (cfcHom_continuous ha), StarAlgHom.map_adjoin]
+  congr
+  simpa using cfcHom_id ha
+
+variable {𝕜}
+
+theorem cfcHom_apply_mem_elemental {a : A} (ha : p a) (f : C(spectrum 𝕜 a, 𝕜)) :
+    cfcHom ha f ∈ elemental 𝕜 a :=
+  range_cfcHom 𝕜 ha ▸ ⟨f, rfl⟩
+
+@[simp]
+theorem cfc_apply_mem_elemental (f : 𝕜 → 𝕜) (a : A) :
+    cfc f a ∈ elemental 𝕜 a :=
+  cfc_cases _ a f (zero_mem _) fun hf ha ↦
+    cfcHom_apply_mem_elemental ha ⟨_, hf.restrict⟩
+
+variable [T2Space A]
+
+open StarSubalgebra elemental in
+protected theorem Commute.cfcHom {a b : A} (ha : p a) (hb₁ : Commute a b)
+    (hb₂ : Commute (star a) b) (f : C(spectrum 𝕜 a, 𝕜)) :
+    Commute (cfcHom ha f) b :=
+  have hb : b ∈ centralizer 𝕜 {a} := by simpa [mem_centralizer_iff] using ⟨hb₁.eq, hb₂.eq⟩
+  le_centralizer_centralizer 𝕜 a (cfcHom_apply_mem_elemental ha f) b (.inl hb) |>.symm
+
+protected theorem IsSelfAdjoint.commute_cfcHom {a b : A} (ha : p a)
+    (ha' : IsSelfAdjoint a) (hb : Commute a b) (f : C(spectrum 𝕜 a, 𝕜)) :
+    Commute (cfcHom ha f) b :=
+  hb.cfcHom ha (ha'.star_eq.symm ▸ hb) f
+
+/-- An element commutes with `cfc f a` if it commutes with both `a` and `star a`.
+
+If the base ring is `ℝ` or `ℝ≥0`, see `Commute.cfc_real` or `Commute.cfc_nnreal` which don't require
+the `Commute (star a) b` hypothesis. -/
+protected theorem Commute.cfc {a b : A} (hb₁ : Commute a b)
```

Comment:
```suggestion
If the base ring is `ℝ` or `ℝ≥0`, see `Commute.cfc_real` or `Commute.cfc_nnreal` which don't require
the `Commute (star a) b` hypothesis. -/
@[grind ←]
protected theorem Commute.cfc {a b : A} (hb₁ : Commute a b)
```
Also, can you add `@[grind =]` to `isSelfAdjoint_iff`? This way we don't need to also tag `IsSelfAdjoint.commute_cfc`.